jsdialog: simplify structure

eszkadev opened the pull request:
This is PR which introduces CSS grid to the jsdialogs instead of old table design.
It should be polished before merge.

Andreas-Kainz commented:
### Align widgets left and right with space between

| before | after |
| -------- | ------ |
| ![Screenshot_20220927_232818](https://user-images.githubusercontent.com/8517736/192639388-e40d0c41-1e04-4d94-a41a-1d4cb36520f9.png) | ![Screenshot_20220927_232739](https://user-images.githubusercontent.com/8517736/192639567-fc26fdad-02e0-478d-bf97-3e43dc1d9c1e.png) |

Andreas-Kainz commented:
### Add padding to grid-cell

`padding: 2px;` was added at notebookbar implementation to `cell` so I also add this 2px at the sidebar grid-cell class.

| before | after |
| -------- | ------ |
| ![Screenshot_20220927_233802](https://user-images.githubusercontent.com/8517736/192640871-29a8c27d-ce6a-44e8-b081-ad1b07c1cce5.png) | ![Screenshot_20220927_233747](https://user-images.githubusercontent.com/8517736/192640895-5aad095f-b041-4bcc-b563-819ca3725c4d.png) |

Andreas-Kainz commented:
### Paragraph section alignment

The spinfields ans margins need some alignment value.

| before | after |
| -------- | ------ |
| ![Screenshot_20220927_235242](https://user-images.githubusercontent.com/8517736/192642807-19e65bbc-c153-41de-8fcf-9a2a79193f63.png) | ![Screenshot_20220927_235227](https://user-images.githubusercontent.com/8517736/192642827-eb12b960-eb1e-4b98-ba5c-cc0c5754c90b.png) |

Andreas-Kainz commented:
### Cell Appearance at calc sidebar appearance

| before | after |
| -------- | ------ |
| ![Screenshot_20220928_000447](https://user-images.githubusercontent.com/8517736/192646247-585f39b4-2d27-480f-af56-5c39d6f4d4ed.png) | ![Screenshot_20220928_000454](https://user-images.githubusercontent.com/8517736/192646264-0030d54f-747b-4d50-b481-72c35693c510.png)

Andreas-Kainz commented:
### Alignment smaller buttons at calc sidebar

| before | after |
| -------- | ------ |
| ![Screenshot_20220928_003258](https://user-images.githubusercontent.com/8517736/192649662-e280c488-56c8-4060-8e02-ec01e78f0607.png) | ![Screenshot_20220928_003316](https://user-images.githubusercontent.com/8517736/192649678-88b4646d-6990-49dc-8819-4ca5843bae6b.png) |

@pedropintosilva do you have an idea how to align the spin boxes and wrap text to the right?

eszkadev commented on browser/css/jssidebar.css:
```diff
@@ -10,6 +10,12 @@
 	vertical-align: middle;
 }
 
+.ui-expander-content.sidebar.expanded > .sidebar .ui-grid-cell {
```
won't be that rule needed not only in expander in sidebar but also in any other grid-cell ?

eszkadev replied:
and in all jsdialogs not only in sidebar?

Andreas-Kainz replied:
as we are editing jssidebar.css from my point of view there should be ONLY sidebar related css rules. If we want them to have also for jsdialogs, than I would move them to jsdialogs. Which would mean more testing, cause it has to work also at dialgoues.

eszkadev commented on browser/css/jssidebar.css:
```diff
@@ -143,6 +153,11 @@
 	justify-content: space-between;
 }
 
+#textorientbox.sidebar button.ui-pushbutton {
```
I have proposal to add these "custom" rules for specific objects at the end of file in one section like:
/* CUSTOMIZED WIDGETS */
so it will be easy to find non-generic rules.

eszkadev commented:
@Andreas-Kainz looks awesome :)

Andreas-Kainz commented:
As I understood at the end of the day EVERYTHING in the sidebar (content area) should be within a `grid-cell `. So that we have to define rules only for that element. This `grid-cell` class is missing for the context related sidebar stuff. I will have a look at that.

However `grid-cell` is fine cause also the notebookbar has also the cell class. My plan is to simplify the css classes maybe to harmonize them with the notebookbar (table) implementation so the structure is similar. Than I will also add some var's for `margin` and `padding` spaces which mean, that we can easy switch between finger/mouse layout.

However first sidebar stuff than harmonization with dialogues and than var's. I want to leave you @eszkadev the message, that I'm interested in support you at this PR.

Andreas-Kainz commented:
@eszkadev as everything has `grid-cell` class. Can we make make a `grid-row` class for each line?

![css classes](https://user-images.githubusercontent.com/8517736/192711993-3109191b-f20b-4a1c-bdc3-38ca82969a7f.png)

Now everythins is a `grid-cell` like the grid-row from the image is also a `grid-cell` so it's difficult to define specific rules.

eszkadev commented:
> @eszkadev as everything has `grid-cell` class. Can we make make a `grid-row` class for each line?
> Now everythins is a `grid-cell` like the grid-row from the image is also a `grid-cell` so it's difficult to define specific rules.

This is what we wanted to do: use CSS grid which automatically does the layouting. We don't want very deep structure. We just removed rows in this PR :)

eszkadev commented:
> However `grid-cell` is fine cause also the notebookbar has also the cell class. My plan is to simplify the css classes maybe to harmonize them with the notebookbar (table) implementation so the structure is similar. Than I will also add some var's for `margin` and `padding` spaces which mean, that we can easy switch between finger/mouse layout.

please do not touch notebookbar in this PR. this change is applied only to sidebar and dialogs now. later we reuse that for notebookbar also

eszkadev commented:
It would be nice to have unified structure based on one grid in every panel:
Here example, Line panel has 3 objects inside, box and 2 grids, alignment in that case is different without setting fixed width...

![alignment](https://user-images.githubusercontent.com/5307369/193027520-89a1c1db-fc0a-4a84-9fe6-8d1ebaeabd0d.png)

Andreas-Kainz commented:
that's the reason for https://bugs.documentfoundation.org/show_bug.cgi?id=148017 to fix this. On the other hand it take ordinary some time to bring the changes from LibO (core) to COOL. However I will start to harmonize the sidebar .ui files. 

Maybe I can add you for the reviews at gerrit and I would like to fix also https://bugs.documentfoundation.org/show_bug.cgi?id=127771 but there I need some help. 

So that we have an standardizised layout at LibO and COOL.

Andreas-Kainz commented:
However do you want to wait for the changes at LibO sidebar or make the fix at LibO and Cool and hope that when the change at LibO was available at Cool it will also work there.

eszkadev commented:
Port of sidebar simplification for core from @Andreas-Kainz to 22.05: https://gerrit.libreoffice.org/c/core/+/141380

eszkadev commented:
> Port of sidebar simplification for core from @Andreas-Kainz to 22.05: https://gerrit.libreoffice.org/c/core/+/141380

Each element goes into the next cell in the grid, I've noticed panels where label should have span=2 because our code assigns only one slot:
![effects](https://user-images.githubusercontent.com/5307369/196094066-39e9f00e-68b0-470e-933c-222ad505014b.png)

Andreas-Kainz commented:
So can I use a prebuild LibO from wget https://github.com/CollaboraOnline/online/releases/download/for-code-assets/core-co-22.05-assets.tar.gz and check this?

eszkadev commented:
> So can I use a prebuild LibO from https://github.com/CollaboraOnline/online/releases/download/for-code-assets/core-co-22.05-assets.tar.gz and check this?

No, I just used this PR and mentioned gerrit patch - locally built. I have some idea how we can automatically generate missing placeholder elements for above issues - have to check. Maybe we don't have to review all the .ui with that :)

eszkadev commented:
> ![effects2](https://user-images.githubusercontent.com/5307369/196094351-eb352e4d-e88c-4edd-9661-329f8cf26903.png)

Above issues are fixed, apart of writer's position and size panel when selected the shape. It's some grid in a grid problem which should be probably fixed in the .ui. TODO: after this rework is accepted.

eszkadev commented:
> > ![effects2](https://user-images.githubusercontent.com/5307369/196094351-eb352e4d-e88c-4edd-9661-329f8cf26903.png)
> 
> Above issues are fixed, apart of writer's position and size panel when selected the shape. It's some grid in a grid problem which should be probably fixed in the .ui. TODO: after this rework is accepted.

fixed in last commit here

Andreas-Kainz commented:
As the 1st column use `justify-content: start;` and the 2nd (or last) column use `justify-content: end;` the simplest solution is use the col number as css class.

![rect10828](https://user-images.githubusercontent.com/8517736/202918318-47257499-d69b-4191-894b-96e2cc0adc1a.png)

I also had a very simple jssidebar.css file, no direct formatting is needed, but I want to discuss first an issue.

#### Issue
Some sidebar widgets didn't has a ui-grid-cell css class and so the widgets also didn't get a ui-grid-col- class. The missing widgets are:
- menubutton
- ui-pushbutton
- ui-listbox-container
- checkbutton
- toolbox
- spinfieldcontainer

can this widgets get an `ui-grid-col-n` class?

eszkadev commented:
> 

@Andreas-Kainz you can do that just using CSS (what is better approach because take into account also cell visibility state):
see: https://stackoverflow.com/a/56315469

eszkadev commented:
I don't like approach from  patch:
[Instead of grid-cell each cell item use grid-col-n so it's easy to se…](https://github.com/CollaboraOnline/online/pull/5256/commits/774f08cfc3faa62ed88fb777fc7c7cc1ccdc6d3f)

What if we will dynamically hide one of the items? It will be broken. Please delegate this work to the browser and CSS language :)

Andreas-Kainz commented:
I know that you can do this with css, but at the .js part you define that the items will be at collumn-n. So if a (hidden) widget is not at the correct column, it has to be fixed at the .js part and the .css part define the widgets width, alignment, ...

I can do the column/row location of widgets, but if you change something at the .js part like a widgets wan't be hidden any more, you have to change the .js part AND the .css part. If we have the column number at the .js part, the .css part is way simplier AND update stuff is super easy.

eszkadev commented:
Problem is that we can at some point dynamically remove one of the elements, or do other modifications -> then numbers will NOT be correct... And browser & CSS always have good results. Also it means - less code. Code is only in CSS. No need to change both JS and CSS.

Also one note here: we should always use "visibility: hidden" in grids instead of "display: none". First one keeps element in the correct place. Second one removes node completely from the calculations.

eszkadev commented:
I've tested without last patch but with merged core patches (2 x simplified .ui files) and it looks great. I don't see any not working widgets. Everything seems to be properly aligned. First column to the left, second to the end. But @Andreas-Kainz  do you want to have content of a second column be aligned to the right? So this screenshot is wrong in your opinion:
![aligned after patches](https://user-images.githubusercontent.com/5307369/203846080-1dc0c8e1-e996-427a-9ed6-74ef43f594f0.png)

Asking to be sure I know what do you want to get here :)

Andreas-Kainz commented:
No change at .js file from me. only changes at jssidebar.css file.

Andreas-Kainz commented:
| writer | calc | impress |
| -------- | ------ | ------------ |
|![writer](https://user-images.githubusercontent.com/8517736/203860938-6dc4dbb6-a025-4ca5-85dc-d2ac340abf15.gif)| ![calc](https://user-images.githubusercontent.com/8517736/203861109-86be9842-0c6b-4fe8-8319-7877eec01a6a.gif)|![impress](https://user-images.githubusercontent.com/8517736/203861157-3a9d2abc-4450-4641-95f3-d5bb9b63204b.gif) |

Andreas-Kainz commented:
all .css classes are for .sidebar so no change anywhere else.

eszkadev commented:
@Andreas-Kainz thank you, now I see what you aligned to the right :)
PR is quite old so I rebased. Unfortunately everything was broken after that so I added some fixes.

eszkadev commented:
![impress](https://user-images.githubusercontent.com/5307369/203909588-d64c8d56-ed61-48ff-afc3-21765a00f40d.png)
noticed one more bug in .ui probably

Andreas-Kainz commented:
Should it be fixed at the .css part or in the .js part?

There is one issue if you are in sidebar charts and select XY chart type.

eszkadev commented:
> 

I think we need to fix .ui file for Impress

Andreas-Kainz commented:
Ok will check on weekend. In general I like the simple structure from now. Also the .css file is way simplier.

Andreas-Kainz commented:
The layout isn't related to the .ui file. it's related to the bullet group which is sometimes visible, sometimes not. The solution would be to add span = 2 when the bullet items are visible.

| first select | after select a text item |
| --------------- | --------------------------------- |
| ![Screenshot_20221126_083027](https://user-images.githubusercontent.com/8517736/204077838-8e736917-8a70-4278-ba41-84a64f9bf3a3.png) | ![Screenshot_20221126_083054](https://user-images.githubusercontent.com/8517736/204077847-d7752346-a103-466d-a330-6395603970d7.png) |

Andreas-Kainz commented:
Here is the second issue

| charts line type | charts XY (Scatter) |
| ---------------------- | --------------------------- |
| ![Screenshot_20221126_085242](https://user-images.githubusercontent.com/8517736/204078491-887a70b9-6a96-4cb4-bdcd-27234126bc3d.png) | ![Screenshot_20221126_085216](https://user-images.githubusercontent.com/8517736/204078499-d7635483-f4af-43a9-8f0f-8dc686d2c7fc.png) |

If I set
```
.sidebar#ChartTypePanel #grid6 {
   display: none;
}
```
I have issues at some other chart type selections.

eszkadev commented:
> > 
> 
> I think we need to fix .ui file for Impress
> 
> > Here is the second issue
> > charts line type 	charts XY (Scatter)
> > ![Screenshot_20221126_085242](https://user-images.githubusercontent.com/8517736/204078491-887a70b9-6a96-4cb4-bdcd-27234126bc3d.png) 	![Screenshot_20221126_085216](https://user-images.githubusercontent.com/8517736/204078499-d7635483-f4af-43a9-8f0f-8dc686d2c7fc.png)
> > If I set
> > ```
> > .sidebar#ChartTypePanel #grid6 {
> >    display: none;
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I have issues at some other chart type selections.
> 
> Good catch. CSS grid requires `visibility: hidden` instead of `display: none` - this should help

I fixed that issue.
